### PR TITLE
Added overloading flexibility to allow extending functionality

### PR DESCRIPTION
**Pull request by diegovdev**

### CODE REVIEW

File: src/wawa-table.ts [hunk 1]
```diff
@@ -213,16 +213,16 @@ export class WawaTable extends LitElement {
     }
 
     protected render(): TemplateResult {
-        return html`${this.renderStyles()}<div style="width:100%;" @scroll=${this.onScroll}>
+        return html`${this.renderStyles()}<div class="table-container" style="width:100%;" @scroll=${this.onScroll}>
```
Comment (diegovdev) [inline src/wawa-table.ts:216]: This allows a subclass to wrap the component in another div and not break query selectors

File: src/wawa-table.ts [hunk 2]
```diff
@@ -213,16 +213,16 @@ export class WawaTable extends LitElement {
...
             <table part="table" style="border-collapse: collapse;width:100%;">
                 <thead part="head">
-                    ${this.headerTemplate}
+                    ${this.composeHeader(this.headerTemplate)}
```
Comment (diegovdev) [inline src/wawa-table.ts:219]: this allows a subclass to change how the header is rendered

File: src/wawa-table.ts [hunk 3]
```diff
@@ -213,16 +213,16 @@ export class WawaTable extends LitElement {
...
                 </thead>
                 <tbody part="body">
                     ${this.startIndex > 0 ? html`<tr style="height:${this.startIndex * this.rowHeight}px"></tr>` : html``}
                     ${repeat(this.items, (i) => i.id, (i, index) => {
                         if (index >= this.startIndex && (this.visibleRows == 0 || index < this.startIndex + this.visibleRows)) {
-                            return html`${i.template}`;
+                            return html`${this.composeRow(index, i)}`;
```
Comment (diegovdev) [inline src/wawa-table.ts]: this allows a subclass to change how a row is rendered

File: src/wawa-table.ts [hunk 4]
```diff
@@ -155,6 +154,7 @@ export class WawaTable extends LitElement {
 
     public connectedCallback(): void {
         super.connectedCallback();
+        this.loadTemplates();
```
Comment (diegovdev) [inline src/wawa-table.ts:157]: moved here from constructor() to have access to members that are not available yet in the constructor

File: src/wawa-table.ts [hunk 5]
```diff
@@ -114,7 +113,7 @@ export class WawaTable extends LitElement {
             this.fetchData(this.pageNumber, this.pageSize).then(async items => {
                 this.moreItems = items.length > 0;
                 for(let i: number = 0; i < items.length; i++) {
-                    this.items.push(new WawaItem(items[i], this));
+                    this.pushItem(items[i]);
```
Comment (diegovdev) [inline src/wawa-table.ts:116]: having a protected pushItem(...) method allows overloading to add functionality when an item is added

**Review by diegovdev — COMMENTED** [review]
All the changes were done to allow a subclass to provide selectable rows automatically, the screenshot of the end result of the subclass is attached here below:

File: src/wawa-table.ts (hunk 3 above)
Comment (matpl) [reply src/wawa-table.ts]: I don't think it's necessary to wrap the composeRow call inside html`` as composeRow already returns a TemplateResult

File: src/wawa-table.ts [hunk 6]
```diff
@@ -234,6 +234,32 @@ export class WawaTable extends LitElement {
...
+    /**
+     * Compose the TemplateResult for the header
+     * @param template The header template
+     */
+    protected composeHeader(template: TemplateResult): TemplateResult {
+        return html`${template}`;
+    }
+
+    /**
+    * Compose the TemplateResult for a row
```
Comment (matpl) [inline src/wawa-table.ts]: Fix indent on the asterisk

**Review by matpl — CHANGES_REQUESTED** [review]

**Pull request merged**